pkgs/by-name/dn/dnspy/deps.nix

```diff
@@ -0,0 +1,85 @@
+  (fetchNuGet { pname = "Microsoft.CodeAnalysis.Workspaces.Common"; version = "4.10.0"; hash = "sha256-5zaaeBacdeqgEv50mJ9g/YN2yhNKvErXyMZsU7coOUM="; })
+  (fetchNuGet { pname = "Microsoft.CSharp"; version = "4.7.0"; hash = "sha256-Enknv2RsFF68lEPdrf5M+BpV1kHoLTVRApKUwuk/pj0="; })
+  (fetchNuGet { pname = "Microsoft.Diagnostics.DbgShim"; version = "8.0.510501"; hash = "sha256-g1QJZvNcCxdS/623XqwkJt3XseTC9sPwWfrQTOjNOxU="; })
+  (fetchNuGet { pname = "Microsoft.Diagnostics.DbgShim.linux-arm"; version = "8.0.510501"; hash = "sha256-Kl27aHMvI2Fyo9iT+SS8N0k+eexd+bf2GjIyxL+2vi4="; })
+  (fetchNuGet { pname = "Microsoft.Diagnostics.DbgShim.linux-arm64"; version = "8.0.510501"; hash = "sha256-2OymQoxpeId2w8YGvrQF+sK3etnl1jItdUv7woBksf4="; })
+  (fetchNuGet { pname = "Microsoft.Diagnostics.DbgShim.linux-musl-arm"; version = "8.0.510501"; hash = "sha256-3AM+MxPRIeQeTQ3bSXzZ07LtjDTAv45u/6xKWN4INDc="; })
+  (fetchNuGet { pname = "Microsoft.Diagnostics.DbgShim.linux-musl-arm64"; version = "8.0.510501"; hash = "sha256-ZBwBrOhRCXEBH0LROqq5pk+D+rSLHxNm+yVYONL4qzY="; })
+  (fetchNuGet { pname = "Microsoft.Diagnostics.DbgShim.linux-musl-x64"; version = "8.0.510501"; hash = "sha256-LxkH4k+XKGELuVCwTAQ5Fg6RP0UibI2VAlkDg4/Pnpg="; })
+  (fetchNuGet { pname = "Microsoft.Diagnostics.DbgShim.linux-x64"; version = "8.0.510501"; hash = "sha256-09e/9r22NdwLKrpXQ4KsUkNI2QRn+E9eKUlVVeynfwY="; })
+  (fetchNuGet { pname = "Microsoft.Diagnostics.DbgShim.osx-arm64"; version = "8.0.510501"; hash = "sha256-MR5+zcf0e16SlNxd0HZ6bOYehUZU2lDK5lr1SJY/62g="; })
+  (fetchNuGet { pname = "Microsoft.Diagnostics.DbgShim.osx-x64"; version = "8.0.510501"; hash = "sha256-Lpgn5rwjf+gH6A56Uf7KyUxTcFXg8YXO4zpAZ/Y46Pg="; })
+  (fetchNuGet { pname = "Microsoft.Diagnostics.DbgShim.win-arm"; version = "8.0.510501"; hash = "sha256-d1sTw6L4rrOW0WulyoEKAkH+bLj6wqd1vsHivVdlbO0="; })
+  (fetchNuGet { pname = "Microsoft.Diagnostics.DbgShim.win-arm64"; version = "8.0.510501"; hash = "sha256-jyhrx+x1jKUgNq0zgyte0f8gExiwjgqduj1fLppbaK4="; })
+  (fetchNuGet { pname = "Microsoft.Diagnostics.DbgShim.win-x64"; version = "8.0.510501"; hash = "sha256-2G/AFIR5/P5NFGzYyC4eQNl+9FZnrbxGjVf0VkQn4FU="; })
+  (fetchNuGet { pname = "Microsoft.Diagnostics.DbgShim.win-x86"; version = "8.0.510501"; hash = "sha256-t+L9LbwakGnx625hVKunpiyX+jV1abcemP5vHiHGRO4="; })
+  (fetchNuGet { pname = "Microsoft.Diagnostics.Runtime"; version = "1.1.142101"; hash = "sha256-6diZT/XBSfF2lpZrmcxvqEm8jRb5uAomoXTFiuZiJc8="; })
+  (fetchNuGet { pname = "Microsoft.DiaSymReader"; version = "2.0.0"; hash = "sha256-8hotZmh8Rb6Q6oD9Meb74SvAdbDo39Y/1m8h43HHjjw="; })
+  (fetchNuGet { pname = "Microsoft.NETCore.App.Host.win-x64"; version = "8.0.8"; hash = "sha256-pkfKvNeb779TUp9jp19peJjCXK3NGpexaFjWwc3dSBo="; })
+  (fetchNuGet { pname = "Microsoft.NETCore.App.Runtime.win-x64"; version = "8.0.8"; hash = "sha256-C0zUiMMAQ9nd1n7PDVoBhCShHzdGI67YFySTpeFH8uE="; })
```

Comment:
@corngood What would be the best way to use windows target packages here (#343837)? Can't it be simply picked up from the `runtimeId` attribute? Or am I supposed to use nixpkgs cross compiling machinery here?

Reply:
> Or am I supposed to use nixpkgs cross compiling machinery here?

In theory building the unwrapped dotnet module cross compiled to windows would be the way to do it.  You'd want the supported platforms to be windows 64 and 32-bit, which would make the target packages for both available at fetch time.

None of that is well tested though, so I wouldn't be surprised if you had trouble.

```
    addNuGetDeps {
      inherit nugetDeps;
      overrideFetchAttrs =
        old:
        lib.optionalAttrs ((args'.runtimeId or null) == null) rec {
          dotnetRuntimeIds = map (system: dotnetCorePackages.systemToDotnetRid system) meta.platforms;
          buildInputs =
            old.buildInputs
            ++ lib.concatLists (lib.attrValues (lib.getAttrs dotnetRuntimeIds dotnet-sdk.targetPackages));
        };
    } args'' finalAttrs
```

`dotnetRuntimeIds` in the above (from `buildDotnetModule`) is where it determines the list of target RIDs needed for restore.  An alternative to actually cross-targeting the whole derivation would be to use `[ "win-x86" "win-x64" ]` there, to match the explicit `runtimeId`.

Reply:
> None of that is well tested though, so I wouldn't be surprised if you had trouble.

I had to make 5a01bbdbfedb27ee0f99932222bb4446c06b5236 but other than that, it seems to work.

> In theory building the unwrapped dotnet module cross compiled to windows would be the way to do it.

In 6f51f2d8efcf005da61d8cbeedd86353ec10448c I just used a cross dotnet sdk instead making the whole unwrapped module cross compile, is that fine?

> An alternative to actually cross-targeting the whole derivation would be to use `[ "win-x86" "win-x64" ]` there, to match the explicit `runtimeId`.

For this package specifically it would actually be either `[ "win-x64" ]`  or  `[ "win-x86"]`, not both. So I think it would work fine if the code you mentioned just added `dotnet-sdk.targetPackages.${args'.runtimeId}` to `buildInputs`, no?
Generally why is `runtimeId` even a thing? Isn't it redundant when the concept of `targetPlatform` exists in nixpkgs? One advantage that I think it has is that unlike gcc, dotnet-sdk can have multiple target platforms installed at a time, so a single `buildDotnetModule` could in theory output binaries for different platforms (say a C# IDE that compiles remote debugger agents for a lot of different platforms as part of its build process). But `dotnetRuntimeIds` isn't even exposed and `runtimeId` is not a list.

Reply:
> For this package specifically it would actually be either [ "win-x64" ] or [ "win-x86"], not both. So I think it would work fine if the code you mentioned just added dotnet-sdk.targetPackages.${args'.runtimeId} to buildInputs, no?

I was thinking you could have the 'unwrapped' derivation have both platforms in meta.platforms, which would allow a single deps.nix to be populated by restoring for both platforms.  In practice it might not matter, if nothing platform specific ends up in deps.nix.

> Generally why is runtimeId even a thing? Isn't it redundant when the concept of targetPlatform exists in nixpkgs?

It probably is redundant now.  I could imagine a list potentially being useful, but I'd want to make sure it interacts in a sensible way with `fetch-deps`.

Reply:
> a single deps.nix to be populated by restoring for both platforms

That feels incorrect though, unless deps.nix had a way to make some dependencies only downloaded on specific rids.

> It probably is redundant now.

So we should pick one and remove the other?

Reply:
> That feels incorrect though, unless deps.nix had a way to make some dependencies only downloaded on specific rids.

That's the way it works for all packages right now.  The deps.nix always contains the dependencies for all supported platforms.  You're right that it would be better to track where packages are actually used, and that's something I'd like us to do eventually (similar to how `targetPackages` was recently added to the sdks). 

> So we should pick one and remove the other?

I'm not sure there's a better option than what's in this PR.  I'm playing around with rearranging this so that is cross compiles `unwrapped` instead of just using the cross sdk, but I think it's going to require more build-support fixes.

Reply:
> I was thinking you could have the 'unwrapped' derivation have both platforms in meta.platforms

I've looked into doing this briefly and looks like `buildDotnetModule` will need a lot more changes for proper cross compilation

Reply:
Just realized you commented before I refreshed

> I'm playing around with rearranging this so that is cross compiles unwrapped

Well... me too. Do you maybe a matrix/discord? :)

Reply:
I'm on matrix @[username]:[username].com.  A nixpkgs-dotnet channel is probably a decent idea.  I mostly gave up lurking on the big nix channels.

Reply:
Yeah, a dotnet channel in the nixos matrix would be a good idea, I wanted to discuss with other dotnet team members the changes in another PR and couldn't find a way to contact them 😅

Reply:
Sorry, I should have mentioned it here, but someone pointed me at an existing one: https://matrix.to/#/#dotnet:nixos.org